qadence/mitigations/zne.py

```diff
@@ -0,0 +1,102 @@
+from qadence.measurements import Measurements
+from qadence.mitigations import Mitigations
+from qadence.types import NoiseModels
+from qadence.utils import Endianness
+
+
+def zne(noise_probas: list, zne_dataset: list) -> float:
+    # Rearrange the dataset by selecting each element in the batches.
+    # TODO: Correct for arbitrary batches.
+    rearranged_dataset = [s[0][0] for s in zne_dataset]
+    # Polynomial fit.
+    p = np.poly1d(np.polyfit(noise_probas, rearranged_dataset, 4))
+    return float(p(0.0))  # Return the zero-noise fitted value.
+
+
+def analog_zne(
+    backend_name: BackendName,
+    circuit: ConvertedCircuit,
+    observable: list[ConvertedObservable] | ConvertedObservable,
+    param_values: dict[str, Tensor] = {},
+    state: Tensor | None = None,
+    protocol: Measurements | None = None,
+    mitigation: Mitigations | None = None,
+    endianness: Endianness = Endianness.BIG,
+) -> Tensor:
+    assert mitigation
+    noise_model = mitigation.options.get("noise_model")
+    if noise_model is None:
+        KeyError(f"A noise model should be choosen from {NoiseModels.list()}. Got {noise_model}.")
+    assert noise_model
+    noise_probas = mitigation.options.get("noise_probas")
+    if noise_probas is None:
+        KeyError(f"A range of noise probabilies should be passed. Got {noise_probas}.")
+    noise_probas = cast(list, noise_probas)
+    backend = backend_factory(backend=BackendName.PULSER, diff_mode=None)
+    backend = cast(Backend, backend)  # Cast the Pulser backend.
+    backend_config = backend.config
+    # Construct the ZNE dataset.
+    zne_dataset = []
+    for noise_proba in noise_probas:
+        # Setting the backend config to account for the noise.
+        if noise_model == NoiseModels.DEPOLARIZING:
+            backend_config.sim_config = SimConfig(noise=noise_model, depolarizing_prob=noise_proba)
+        elif noise_model == NoiseModels.DEPHASING:
+            backend_config.sim_config = SimConfig(noise=noise_model, dephasing_prob=noise_proba)
+        # Get density matrices in the noisy case.
+        density_matrices = backend.run_noisy(
+            circuit, param_values=param_values, state=state, endianness=endianness
+        )
+        # Set the sim_config state back to original for any further use.
+        # self.config.sim_config = None
+        # Convert observables to Numpy types compatible with QuTip simulations.
+        # Matrices are flipped to match QuTip conventions.
+        observables_np = [np.flip(block_to_tensor(obs.original).numpy()) for obs in observable]
+        # Get expectation values at the end of the time serie [0,t]
+        # at intervals of the sampling rate.
+        zne_dataset.append(
+            [[dm.expect(obs)[0][-1] for obs in observables_np] for dm in density_matrices]
+        )
+    # Zero-noise extrapolate.
+    exp_val = zne(noise_probas=noise_probas, zne_dataset=zne_dataset)
+    return torch.tensor([exp_val])
+
```

Comment:
The `analog_zne` looks like a very basic implementation of the zero noise extrapolation scheme. Isn't there any unfolding happening? I see only the final fit (extrapolation) to zero noise.

Reply:
#181